### ISSUE TITLE
Design basic action management module(s)

### ISSUE THREAD

**Issue by borjasotomayor** [uchicago-cs/chiventure#7]
Write header files that describe the data structures and functions necessary to create and manipulate actions (note: there is a separate issue for coming up with the specification of actions: #6)

**Comment by MTUC08** [comment]
After the meeting today, Yoon and I met up to discuss structures and functions for action management and came up with the following as a preliminary structure:

We will have an enum for actions, including all of the possible actions that the other half of our group has decided upon. These actions will be of three types: act_npc, act_world, and act_inventory, which act upon npc, the world, and inventory respectively. The type will also be stored as an enum.

Our action struct will contain the actions enum, the action_type enum, and a success and failure message. 

We assume that object structs in game state will contain an int array of allowed_actions that correspond to our actions enum, and int telling us how long that array is, and the size of the object (so we know if we can add it to inventory and update the inventory as needed). 

Given that assumption, we would expect that game state would receive an action and object from CLI, call our action_init/new functions to create an action_struct, then depending on what type of action it is, call either act_inventory, act_npc, or act_world.  These functions will then use the enum action as a tag, and with a switch statement, redirect to their respective action functions to carry out the action.

**Comment by MTUC08** [comment]
After conversing in slack about which actions to take out (go to/move to) and which to condense into one (consume), we pushed the final product into the action_management branch under include. We have an action management.h file that includes the structs and functions we need to implement actions, and a structs.h file that contains some of our preliminary requirements for structs from game state. While a number of additional components can be added to the structs, for our preliminary action modules, we need to have those structs with those components at a minimum.

**Comment by yoonsjeong** [comment]
The current list of "parent actions" we have are the following
{OPEN,CLOSE,PUSH,PULL,EXAMINE,TURN_ON,TURN_OFF,TALK_TO,GIVE,GO,TAKE,DROP,CONSUME,USE_ON}

These actions are how commands like "open door" or "close door" will be represented in the actual code. That means that "action synonyms" will be represented using these as well. For example, "walk EAST" and "go EAST" would be identical for the purpose of altering the game state, and therefore the enumerated variable "GO" would be used to represent both.

Here is our tentative list of synonyms (please feel free to critique/suggest)

EXAMINE: examine, look at
TALK_TO: talk to, talk
GIVE: give to, give
GO: walk, go
TAKE: pick up, take
DROP: drop
CONSUME: eat, drink, use

**Comment by yoonsjeong** [comment]
In our module, we use a number of "assumed structs" which the Game State team would be responsible for. Those structs include the npc_struct, room_struct, object_struct, and player_struct.

Within those structs we assume player_struct contains the information for the inventory, that the object_struct contains the actions allowed to be performed on that action, and that the room_struct contains the objects and npcs capable of being acted upon.

Here is the file with all of the structs and details:
https://github.com/uchicago-cs/chiventure/blob/action_management/src/action_management/include/structs.h

**Comment by MTUC08** [comment]
@mpcsh @borjasotomayor Do you mind confirming that this issue is good to close? Our action structs and functions are contained in: https://github.com/uchicago-cs/chiventure/blob/action_management/src/action_management/include/actionmanagement.h, and our requirements for components in other structs is in: https://github.com/uchicago-cs/chiventure/blob/action_management/src/action_management/include/structs.h

**Comment by borjasotomayor** [comment]
Yes, please go ahead and close it.

**Comment by MTUC08** [comment]
Just wanted to give an update:

After creating a pull request and seeing everyone's suggestions, I removed the success and failure messages from the action struct as they can be crafted within each actions implementation, putting less on the heap. I also removed those and the action_type enum from the inputs for the action_new and action_init functions as the action type can be determined from the action input within action_init. Finally, I fixed our description for the map recommendation in our structs.h file. Thanks everyone!
